tests/foreman/cli/test_leapp_client.py

```diff
@@ -0,0 +1,257 @@
+        all_repos.append(rh_repo)
+        # sync repo
+        rh_repo.sync(timeout=1800)
+    cv.repository = all_repos
+    cv = cv.update(['repository'])
+    # Publish, promote content view to lce
+    cv.publish()
+    cvv = cv.read().version[0]
+    cvv.promote(data={'environment_ids': lc_env.id, 'force': True})
+    cv = cv.read()
+    # Create activation key
+    ak = create_activation_key(target_sat, cv, lc_env, organization)
+
+    # 3. Register Host
+    register_host_with_satellite(target_sat, custom_host, organization, ak)
+
+    # 4. Verify target rhel version repositories has enabled on Satellite Server
+    verify_target_repo_on_satellite(target_sat, cv, organization, lc_env, target_rhel='rhel-9')
+
+    # 5. Update all packages and install Leapp utility
+    # Preupgrade conditions and check
+    rhel_old_ver = custom_host.run('cat /etc/redhat-release')
+    precondition_check_upgrade_and_install_leapp_tool(custom_host, custom_host.deploy_rhel_version)
+
+    # Fixing inhibitors to avoid hard stop of Leapp tool execution
+    fix_inhibitors(custom_host, custom_host.deploy_rhel_version)
+
+    # 6. Run LEAPP-PREUPGRADE Job Template-
+    template_id = (
+        target_sat.api.JobTemplate()
+        .search(query={'search': 'name="Run preupgrade via Leapp"'})[0]
+        .id
+    )
+    job = target_sat.api.JobInvocation().run(
+        synchronous=False,
+        data={
+            'job_template_id': template_id,
+            'targeting_type': 'static_query',
+            'search_query': f'name = {custom_host.hostname}',
+        },
+    )
+    target_sat.wait_for_tasks(
+        f'resource_type = JobInvocation and resource_id = {job["id"]}', poll_timeout=1800
+    )
+    result = target_sat.api.JobInvocation(id=job['id']).read()
+    assert result.succeeded == 1
+    # Run LEAPP-UPGRADE Job Template-
+    template_id = (
+        target_sat.api.JobTemplate().search(query={'search': 'name="Run upgrade via Leapp"'})[0].id
+    )
+    job = target_sat.api.JobInvocation().run(
+        synchronous=False,
+        data={
+            'job_template_id': template_id,
+            'targeting_type': 'static_query',
+            'search_query': f'name = {custom_host.hostname}',
+            'inputs': {'Reboot': 'true'},
+        },
+    )
+    target_sat.wait_for_tasks(
+        f'resource_type = JobInvocation and resource_id = {job["id"]}', poll_timeout=1800
+    )
+    result = target_sat.api.JobInvocation(id=job['id']).read()
+    assert result.succeeded == 1
+    # Wait for the host to be rebooted and SSH daemon to be started.
+    try:
+        wait_for(
+            custom_host.connect,
+            fail_condition=lambda res: res is not None,
+            handle_exception=True,
+            raise_original=True,
+            timeout=180,
+            delay=1,
+        )
+    except ConnectionRefusedError:
+        raise ConnectionRefusedError('Timed out waiting for SSH daemon to start on the host')
```

Comment:
With the patch from #12144, you can use a one-line function to do what you are doing here on 10 lines. You could use the `wait_for_connection` method.